### ISSUE TITLE
ActiveStorage changes in nested model not detected during update

### ISSUE THREAD

**Issue by camiloforero** [rails/rails#37701]
### Steps to reproduce
Having two models:
```ruby
class Parent < ApplicationRecord
  has_many: :children
  accepts_nested_attributes_for :children, allow_destroy: true
end
```
and
```ruby
class Child < ApplicationRecord
  # Assume this Child class has a "name" column in the DB
  has_one_attached :example_file
end
```
And updating the Child model through a very vanilla Parent's controller:
```ruby
  def update
    respond_to do |format|
      if @parent.update(parent_params)
        format.html { redirect_to @parent, notice: 'Parent was successfully updated.' }
        format.json { render :show, status: :ok, location: @parent }
      else
        format.html { render :edit }
        format.json { render json: @parent.errors, status: :unprocessable_entity }
      end
    end
  end

  def parent_params
      params.require(:parent).permit(
        :name, children_attributes: [:id, :name, :example_file, :_destroy],
      )
    end
```
1. Create a new Parent with some Children in it, and save
3. Edit the parent, change the `example_file` and the `name` of a `Child`, save, check the results
2. Edit the parent, change **only the `example_file` of a `Child`**, save, check the results

### Expected behavior
1. In both cases the `example_file` should be updated. (in the first one the `name` is updated as well)

### Actual behavior
The first case will work as expected. But in the second case, the `Child` with the new `example_file` will not be detected as changed, and the uploaded file will not be saved

### System configuration
**Rails version**: 6.0.0

**Ruby version**: 2.6.2

**Comment by pixeltrix** [comment]
@camiloforero can you try with `has_many :children, autosave: true` - by default only new records are saved, not updated ones:

```
:autosave
    If true, always save the associated objects or destroy them if marked
    for destruction, when saving the parent object. If false, never save or
    destroy the associated objects. By default, only save associated
    objects that are new records. This option is implemented as a before_save
    callback. Because callbacks are run in the order they are defined,
    associated objects may need to be explicitly saved in any user-defined
    before_save callbacks.
```

**Comment by camiloforero** [comment]
@pixeltrix I just tried and it didn't work either.

Also, keep in mind that Rails will still detect the changes in the nested model when I update a regular database field, without needing to set the `autosave` option. So there must be something off with the way Rails is doing this change detection

**Comment by pixeltrix** [comment]
Hi @camiloforero - thanks for trying that.

The `has_one_attached` uses a `has_one` and `has_one :through` association underneath the hood and assignments to them happen immediately so unsure what's happening - can you create a single file example or a repo containing a simple rails app that demonstrates the problem, thanks.

**Comment by willtcarey** [comment]
I wanted to chime in and say I'm hitting this same issue. I'm the same setup as @camiloforero: A Parent record which `has_many` children and `accepts_nested_attributes_for` children. The suggestion earlier in this thread to try putting `autosave: true` on the relationship isn't needed because having the `accepts_nested_attributes_for` call turns the association into an autosave association.

I was able to trace it back to where the autosave association [checks all the collection records to see if they need to be saved](https://github.com/rails/rails/blob/4a17d3e045ef059d5c82d1ade4f7eae37bc1e485/activerecord/lib/active_record/autosave_association.rb#L284) and the Child record returns false even though the ActiveStorage attachment has changed (checked with `example_file.changed_for_autosave?` on the Child record).

It seems as though the record is not properly checking ActiveStorage attachments for their changes.

My temporary workaround has been to override the `changed_for_autosave?` method in my child model with
```
class Child < ApplicationRecord
  ...
  
  def changed_for_autosave?
    super || example_file.changed_for_autosave?
  end
end
```

I created a repository to reproduce the issue [here](https://github.com/willtcarey/active-storage-change-bug).

Let me know if you need any other help in reproducing this issue!

**Comment by pixeltrix** [comment]
@willtcarey thanks for the repo - helped track it down.

First thing I thought was the associations that Active Storage creates aren't defined with `autosave: true`, however turns out this is irrelevant because the actual association isn't updated until an `after_save` callback so `changed?` on the child wouldn't return true anyway. Looks to be a fairly simple fix to override `changed_for_autosave?` in `ActiveStorage::Attached::Model` to check for attachment changes, e.g:

```
def changed_for_autosave?() #:nodoc:
  super || !attachment_changes.empty?
end
```

Anyone want to have a go a fixing it?

**Comment by willtcarey** [comment]
@pixeltrix I created a pull request for a fix. This is my first pull request on rails so I hope I got everything right. I wrote a test to go along with the fix but let me know if I need anything else!

**Comment by willtcarey** [comment]
@pixeltrix Sorry to bother, but I'm checking to see if you need anything else from me to push this through?

**Comment by otaviogaiao** [comment]
> def changed_for_autosave?
>     super || example_file.changed_for_autosave?
>   end

Thank you very much. Your workaround works perfectly. I was having the same issue.

**Comment by r-obeen** [comment]
> class Child < ApplicationRecord
>   ...
>   
>   def changed_for_autosave?
>     super || example_file.changed_for_autosave?
>   end
> end
> ```

This did the trick for me.

Anyone know what version for Rails will have a fix ? And when will it be released ?

**Comment by stevejc** [comment]
I'm running into a similar problem on any field in the nested attributes, not just an attached file. I can saved_changes on create, but nothing on the update. I upgraded to 6.0.3.1 and still have the issue. I'm assuming this fix is only for attached files. Should I submit a separate issue?

**Comment by jonathanhefner** [comment]
@stevejc 6.0.3.1 was a security patch release.  It only included CVE fixes.  The fix from #37786 will be included in Rails 6.1 (since it was committed to `master` branch) and Rails 6.0.4 (since it was backported to `6-0-stable` branch).
